src/components/ingestion/AbstractIngestionEngine.js

```diff
@@ -1,66 +1,120 @@
 import { spawn } from 'child_process'
+import {
+	pipeline,
+	Writable,
+	Transform,
+} from 'stream'
 import { Payload } from '@tvkitchen/base-classes'
 import { dataTypes } from '@tvkitchen/base-constants'
 import {
 	AbstractInstantiationError,
 	NotImplementedError,
 } from '@tvkitchen/base-errors'
+import { TSDemuxer } from 'ts-demuxer'
+
 import logger from '%src/lib/logger'
 import kafka from '%src/lib/kafka'
+import { tsToSeconds } from '%src/lib/utils/mpegts'
 
 class AbstractIngestionEngine {
 	constructor() {
 		if (this.constructor === AbstractIngestionEngine) {
 			throw new AbstractInstantiationError(this.constructor.name)
 		}
 		this.producer = kafka.producer()
+		this.mpegtsDemuxer = new TSDemuxer(this.processDemuxedPacket)
+		this.mostRecentDemuxedPacket = null
 	}
 
 	/**
-	 * Ingests data into the TV Kitchen pipeline.
+	 * Ingests an mpegts payload into the TV Kitchen pipeline.
 	 *
-	 * @param  {Buffer} data The buffer of data to be ingested.
-	 * @return {Boolean}     Whether or not the buffer was successfully ingested.
+	 * @param  {Payload} data The payload to be ingested.
+	 * @return {Boolean}      Whether or not the payload was successfully ingested.
 	 */
-	ingestData = async (data) => {
-		const payload = new Payload()
-		payload.data = data
-		await this.producer.send({
-			topic: dataTypes.STREAM.DATA,
-			messages: [{
-				value: data,
-			}],
-		})
-		return true
+	ingestPayloadStream = Writable({
+		objectMode: true,
+		write: async (payload, enc, done) => {
+			if (!(payload instanceof Payload)) {
+				done(new Error('ingestPayloadStream received non-Payload data'))
+			} else {
+				done(this.producer.send({
+					topic: dataTypes.STREAM.CONTAINER,
+					messages: [{
+						value: Payload.serialize(payload),
+					}],
+				}))
+			}
+		},
+	})
+
+	/**
+	 * Updates ingestion state based on the latest demuxed packet data.
+	 *
+	 * This method is called by our mpegts demuxer, and allows the ingestion engine to track
+	 * the most recent demuxed packet.
+	 *
+	 * @param  {Packet} packet The latest TSDemuxer Packet object.
+	 */
+	processDemuxedPacket = (packet) => {
+		this.mostRecentDemuxedPacket = packet
 	}
 
 	/**
-	 * Continuously ingest of a stream of data.
+	 * Returns the most recent cohearant stream packet processed by the ingestion engine.
+	 * This packet is lower level than the mpegts container, and represents an audio or video
+	 * packet demuxed from the mpegts stream.
+	 *
+	 * @return {Packet} The most recent Packet object extracted by TSDemuxer.
+	 */
+	getMostRecentDemuxedPacket = () => this.mostRecentDemuxedPacket
+
+	/**
+	 * Process a new chunk of data from an mpegts stream. The chunks passed to this
+	 * function should be presented sequentially, and combine to form a cohearant mpegts
+	 * data stream, but they can be of arbitray size.
```

Comment:
```suggestion
	 * data stream, but they can be of arbitrary size.
```

Reply:
Woah what even is this interface option...

Reply:
I found it when Googling “how to suggest changes to a pull request.” How to make them under item 6 [here](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/commenting-on-a-pull-request#adding-line-comments-to-a-pull-request) and how to incorporate them [here](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/incorporating-feedback-in-your-pull-request#applying-suggested-changes).